runtime/compiler/optimizer/J9Simplifier.cpp

```diff
@@ -479,28 +479,73 @@ TR::Node *
-       (node->getReferenceCount() == 1) &&
-       performTransformation(comp(), "%sRemoved dead BigDecimal/BigInteger call node [" POINTER_PRINTF_FORMAT "]\n", optDetailString(), node))
+       (node->getSymbol()->isResolvedMethod()))
       {
-      TR::Node *firstChild = node->getFirstChild();
-      anchorChildren(node, _curTree);
+      bool isSimplifiableMethod = false;
+      bool isIntOrDecimalNonDivArithOp = false;
 
-      firstChild->incReferenceCount();
+      switch (node->getSymbol()->getResolvedMethodSymbol()->getRecognizedMethod())
+         {
+         case TR::java_math_BigDecimal_valueOf:
+            {
+            isSimplifiableMethod = true;
+            break;
+            }
+
+         case TR::java_math_BigDecimal_add:
+         case TR::java_math_BigDecimal_subtract:
+         case TR::java_math_BigDecimal_multiply:
+         case TR::java_math_BigInteger_add:
+         case TR::java_math_BigInteger_subtract:
+         case TR::java_math_BigInteger_multiply:
+            {
+            isSimplifiableMethod = true;
+            isIntOrDecimalNonDivArithOp = true;
+            break;
+            }
+
+         default:
+            {
+            isSimplifiableMethod = false;
+            isIntOrDecimalNonDivArithOp = false;
+            break;
+            }
+         }
+
+      if (isSimplifiableMethod && (node->getReferenceCount() == 1) &&
+          performTransformation(comp(),
+                isIntOrDecimalNonDivArithOp ?
+                      ("%sReplaced dead BigDecimal/BigInteger call node [" POINTER_PRINTF_FORMAT "] with NULLCHK of argument\n") :
+                      ("%sRemoved dead BigDecimal/BigInteger call node [" POINTER_PRINTF_FORMAT "]\n"),
+                optDetailString(), node))
+         {
+         TR::Node *firstChild = node->getFirstChild();
+         anchorChildren(node, _curTree);
 
-      int i;
-      for (i = 0; i < node->getNumChildren(); i++)
-         node->getChild(i)->recursivelyDecReferenceCount();
+         firstChild->incReferenceCount();
 
-      TR::Node::recreate(node, TR::PassThrough);
-      node->setNumChildren(1);
+         // Need to preserve NULLCHK for argument to
+         // Big{Integer|Decimal}.{add|multiply|subtract}
+         if (isIntOrDecimalNonDivArithOp)
+            {
+            TR::Node *secondChild = node->getSecondChild();
+            TR::SymbolReference * const nullCheckSR =
+                  comp()->getSymRefTab()->findOrCreateNullCheckSymbolRef(
+                                                comp()->getMethodSymbol());
+            TR::TreeTop::create(comp(), _curTree,
+                  TR::Node::createWithSymRef(node, TR::NULLCHK, 1,
+                        TR::Node::create(node, TR::PassThrough, 1, secondChild),
+                                         nullCheckSR));
+            _alteredBlock = true;
+            }
+
+         int i;
```

Comment:
can we please move the declaration of i into the for loop to limit its scope since you aren't using it outside the loop

Reply:
Fixed in https://github.com/eclipse/openj9/commit/ed97fb6168f0220071a8694fe1b19cc919d3af73